onnxruntime/contrib_ops/cpu/tokenizer.cc

```diff
@@ -365,77 +192,85 @@ Status Tokenizer::CharTokenize(OpKernelContext* ctx, size_t N, size_t C,
+    if (!utf8_validate(reinterpret_cast<const unsigned char*>(s.data()), s.size(),
+                       utf8_chars)) {
       return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
-                    "Invalid utf8 chars in the input: " + s);
+                    "Input string contains invalid utf8 chars: " + s);
     }
 
-    std::set<Match> matches;
-    const wchar_t* ws = wstr.c_str();
-    size_t len_remaining = wstr.length();
-    size_t offset = 0;
-    while (len_remaining > 0) {
-      const auto* val = search_data_->tst_.get(ws, len_remaining);
-      if (val != nullptr) {
-        auto p = matches.insert({val->priority_, offset, val->w_len});
-        while (!p.second && val->priority_ < p.first->priority_) {
-          // if overlapping matches of the same pattern(priority), then
-          // the earlier match naturally wins
-          matches.erase(p.first);
-          p = matches.insert({val->priority_, offset, val->w_len});
+    std::vector<StringPiece> row{s};
+
+    for (const auto& sep : separators_) {
+      std::vector<StringPiece> tokens;
+      for (const auto& text : row) {
+        const auto end_pos = text.length();
+        size_t start_pos = 0;
+        StringPiece submatch;
+
+        bool match = true;
+        while (match) {
+          match = sep->Match(text, start_pos, end_pos, anchor, &submatch, 1);
+          if (match) {
+            // Record  pos/len
+            assert(submatch.data() != nullptr);
+            size_t match_pos = submatch.data() - text.data();
+            assert(match_pos >= start_pos);
+            auto token_len = match_pos - start_pos;
+            utf8_chars = 0;
+            bool valid = utf8_len(reinterpret_cast<const unsigned char*>(text.data() + start_pos),
+                                  token_len, utf8_chars);
+            if (!valid) {
+              return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
+                            "Match contains invalid utf8 chars: " + submatch.as_string());
+            }
+            if (utf8_chars >= size_t(mincharnum_)) {
+              tokens.emplace_back(text.data() + start_pos, token_len);
+            }
+            // Update starting position
+            // Guard against empty string match
+            auto match_len = submatch.length();
+            if (match_len > 0) {
+              start_pos = match_pos + match_len;
+            } else {
+              size_t bytes = 0;
+              utf8_bytes(*submatch.data(), bytes);
+              start_pos = match_pos + bytes;
+            }
+          } else {
+            // record trailing token
+            auto trailing_len = end_pos - start_pos;
+            utf8_chars = 0;
+            utf8_len(reinterpret_cast<const unsigned char*>(text.data() + start_pos),
+                     trailing_len, utf8_chars);
+            if (utf8_chars >= size_t(mincharnum_)) {
+              tokens.emplace_back(text.data() + start_pos, trailing_len);
```

Comment:
No your code. But this extracting and copying thing makes me wonder if string should be linked list in ORT.

Reply:
Linked list would require many small allocations so I generally avoid. There is no copying here and this is append only.

---
In reply to: [276862110](https://github.com/Microsoft/onnxruntime/pull/857#discussion_r276862110) [](ancestors = 276862110)

Reply:
Clarified that there is no copying done.

---
In reply to: [277049667](https://github.com/Microsoft/onnxruntime/pull/857#discussion_r277049667) [](ancestors = 277049667,276862110)